### PR TITLE
Add interface for model flags

### PR DESCRIPTION
**Pull request by TheNormalnij**
**API:**
```lua
bool engineSetModelFlags( int modelID, int flags [, bool bIdeFlags] )
int engineGetModelFlags( int modelID )
bool engineResetModelFlags( int modelID )
bool engineGetModelFlag( int modelID, string flagName )
bool engineSetModelFlag( int modelID, string flagName, bool state )
```

Flags:
```
is_road
draw_last
addictive
ignore_lighting
no_zbuffer_write
dont_receive_shadows
is_glass_type_1
is_glass_type_2
is_garage_door
is_damagable
is_tree
is_palm
does_not_collide_with_flyer
is_tag
disable_backface_culling
is_breakable_statue
is_crane
```

[Flags documentation](https://gtamods.com/wiki/Item_Definition)

**Important**
`engineGetModelFlags` returns internal GTA flags.
You can transform IDE flags into GTA flags with 3 argument in `engineSetModelFlags`.

### CODE REVIEW

File: Client/game_sa/CModelInfoSA.h [hunk 1]
```diff
@@ -36,7 +36,9 @@ class CPedModelInfoSAInterface;
 #define     ARRAY_ModelInfo                 0xA9B0C8
 #define     CLASS_CText                     0xC1B340
 #define     FUNC_CText_Get                  0x6A0050
-#define     FUNC_GetModelFlags              0x4044E0
+#define     FUNC_SetBaseModelFlags          0x5B3AD0
+#define     FUNC_SetAtimicModelFlags        0x5B3B20
```
Comment (StrixG) [inline Client/game_sa/CModelInfoSA.h]:
```suggestion
#define     FUNC_SetAtomicModelFlags        0x5B3B20
```

File: Client/game_sa/CModelInfoSA.cpp [hunk 2]
```diff
@@ -482,21 +483,67 @@ BOOL CModelInfoSA::DoIsLoaded()
...
+    if (!m_pInterface)
+        return;
+
+    // Save default value if not done yet
+    if (!MapContains(ms_ModelDefaultFlagsMap, m_dwModelID))
+        MapSet(ms_ModelDefaultFlagsMap, m_dwModelID, m_pInterface->usFlags);
+
+    m_pInterface->usFlags = 0xC0;
+    typedef char(__cdecl * Function_SetAtomicModelFlags)(CBaseModelInfoSAInterface * pInterface, unsigned int uiFlags);
+    Function_SetAtomicModelFlags setAtomicModelFlags = (Function_SetAtomicModelFlags)(FUNC_SetAtimicModelFlags);
```
Comment (StrixG) [inline Client/game_sa/CModelInfoSA.cpp]:
```suggestion
    Function_SetAtomicModelFlags setAtomicModelFlags = (Function_SetAtomicModelFlags)(FUNC_SetAtomicModelFlags);
```

**Comment by TheNormalnij** [comment]
Can we merge this?

**Comment by Woovie** [comment]
I don't feel it's ready for merge. I tested this codebase while @qaisjp watched as well. The API being provided for get/set is not adequate. It's difficult to use, confusing, and generally not user friendly. We don't have a recommendation/solution for this, but something should change to make it easier for people to use.

**Comment by dmi7ry** [comment]
> I don't feel it's ready for merge. I tested this codebase while @qaisjp watched as well. The API being provided for get/set is not adequate. It's difficult to use, confusing, and generally not user friendly. We don't have a recommendation/solution for this, but something should change to make it easier for people to use.

Do you mean bitwise operations? If so then it would be possible to add functions `engineSetModelFlag(modelID, "flag", state)` and `engineGetModelFlag(modelID, "flag")`. Like `engineSetModelFlag(modelID, "isRoad", true)`.

File: Client/game_sa/CModelInfoSA.h [hunk 3]
```diff
@@ -291,6 +302,7 @@ class CModelInfoSA : public CModelInfo
     RpClump*                                                                     m_pCustomClump;
     static std::map<unsigned short, int>                                         ms_RestreamTxdIDMap;
     static std::map<DWORD, float>                                                ms_ModelDefaultLodDistanceMap;
+    static std::map<DWORD, unsigned short>                                       ms_ModelDefaultFlagsMap;
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.h:305]: Why not combine these two into one?
Also, I feel like an `std::map` has a bigger memory overhead(3 pointers / key value pair), than just a large fixed size array(which also would be a lot faster, and would be less of a hassle).
What do others think?
Perhaps merge it like this, and let me open a separate PR merging them into an `std::array`?

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 4]
```diff
@@ -2017,7 +2024,58 @@ int CLuaEngineDefs::EngineRestoreObjectGroupPhysicalProperties(lua_State* luaVM)
...
+            if (bIdeFlags)
+                pModelInfo->SetIdeFlags(uiFlags);
+            else
+                pModelInfo->SetFlags(uiFlags);
+
+            return true;
+        }
+    }
+
+    throw std::invalid_argument("Expected a valid model ID in range [0-19999] at argument 1");
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp]:
```cpp
if (uiModelID >= 20000) 
    throw std::invalid_argument("Expected a valid model ID in range [0-19999] at argument 1");
``` 
would be better imho. But im just nitpicky.

**Review by Pirulax — APPROVED** [review]

**Comment by TheNormalnij** [comment]
[Table ](https://docs.google.com/spreadsheets/d/1sDE8lI1LiuwdzqzRblIeq-N0gp_oPJjP68juVtD0y5U/edit?usp=sharing) with all model flags

**Comment by Woovie** [comment]
> [Table ](https://docs.google.com/spreadsheets/d/1sDE8lI1LiuwdzqzRblIeq-N0gp_oPJjP68juVtD0y5U/edit?usp=sharing) with all model flags

Private sheet

**Comment by TheNormalnij** [comment]
Fixed

**Comment by TheNormalnij** [comment]
Ready

File: Client/game_sa/CModelInfoSA.cpp [hunk 5]
```diff
@@ -19,12 +19,50 @@ CBaseModelInfoSAInterface** ppModelInfo = (CBaseModelInfoSAInterface**)ARRAY_Mod
 
 std::map<unsigned short, int>                                         CModelInfoSA::ms_RestreamTxdIDMap;
 std::map<DWORD, float>                                                CModelInfoSA::ms_ModelDefaultLodDistanceMap;
+std::map<DWORD, unsigned short>                                       CModelInfoSA::ms_ModelDefaultFlagsMap;
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.cpp:31]: `std::map` has a huge overhead for such small values, so having them all in a single unified map would actually reduce both CPU and RAM usage.
For this PR its okay, but Im going to make a separate PR to unify these maps.

File: Client/game_sa/CModelInfoSA.cpp [hunk 6]
```diff
@@ -19,12 +19,50 @@ CBaseModelInfoSAInterface** ppModelInfo = (CBaseModelInfoSAInterface**)ARRAY_Mod
...
+        char bFlag20 : 1;
+        char bIsTag : 1;
+        char bDisableBackfaceCilling : 1;
+        char bIsBreakableStatue : 1;
+        char bFlag24 : 1;
+
+        char cPad : 8;
+    };
+    unsigned int uiFlags;
+};
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.cpp]: FYI from the [cppreference](https://en.cppreference.com/w/cpp/language/union):
> It's undefined behavior to read from the member of the union that wasn't most recently written. Many compilers implement, as a non-standard language extension, the ability to read inactive members of a union.

I know we are doing it this way since forever, but still, relying on UB is a very bad idea.
Altough, I have no idea what would be the correct way to do this.

Sorry for being nitpicky.

File: Client/game_sa/CModelInfoSA.cpp [hunk 7]
```diff
@@ -483,21 +521,278 @@ BOOL CModelInfoSA::DoIsLoaded()
...
 {
-    DWORD dwFunc = FUNC_GetModelFlags;
-    DWORD ModelID = m_dwModelID;
-    BYTE  bFlags = 0;
-    _asm
+    m_pInterface = ppModelInfo[m_dwModelID];
+    if (m_pInterface)
+        return m_pInterface->usFlags;
+    return 0;
+}
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.cpp]: Is this function supposed to be called with non-initialized models?
Because If not, then we shouldn't be returning `0` here.

File: Client/game_sa/CModelInfoSA.cpp [hunk 8]
```diff
@@ -483,21 +521,278 @@ BOOL CModelInfoSA::DoIsLoaded()
...
+
+unsigned short CModelInfoSA::GetOriginalFlags()
+{
+    if (MapContains(ms_ModelDefaultFlagsMap, m_dwModelID))
+        return MapGet(ms_ModelDefaultFlagsMap, m_dwModelID);
+
+    m_pInterface = ppModelInfo[m_dwModelID];
+    if (!m_pInterface)
+        return 0;
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.cpp]: Same

File: Client/game_sa/CModelInfoSA.cpp [hunk 9]
```diff
@@ -483,21 +521,278 @@ BOOL CModelInfoSA::DoIsLoaded()
...
+
+            if (ideFlags.bIsBreakableStatue)
+                m_pInterface->eSpecialModelType = eModelSpecialType::BREAKABLE_STATUE;
+
+            if (ideFlags.bFlag24)
+                m_pInterface->eSpecialModelType = eModelSpecialType::UNKNOW_1;
+
+            // Added
+            if (ideFlags.bFlag20)
+                m_pInterface->eSpecialModelType = eModelSpecialType::CRANE;
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.cpp]: It seems to me that you should use `else if` here, because (it seems like to me) `m_pInterface->eSpecialModelType` should only be set once?
If so, please use `else if`.

File: Client/game_sa/CModelInfoSA.cpp [hunk 10]
```diff
@@ -483,21 +521,278 @@ BOOL CModelInfoSA::DoIsLoaded()
...
+            return false;
+    }
+}
+
+void CModelInfoSA::SetModelSpecialType(eModelSpecialType eType, bool bState)
+{
+    if (bState)
+        m_pInterface->eSpecialModelType = eType;
+    else if (m_pInterface->eSpecialModelType == eType)
+        m_pInterface->eSpecialModelType = eModelSpecialType::NONE;
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.cpp:718]: I think this is more straightforward: 
```cpp
if (bState)
        m_pInterface->eSpecialModelType = eType;
    else 
          if (m_pInterface->eSpecialModelType == eType)
                m_pInterface->eSpecialModelType = eModelSpecialType::NONE;
```

File: Client/sdk/game/CModelInfo.h [hunk 11]
```diff
@@ -28,6 +28,27 @@ class CBoundingBox
...
+    IS_DAMAGABLE,
+    IS_TREE,
+    IS_PALM,
+    DOES_NOT_COLLIDE_WITH_FLYER,
+    IS_TAG,
+    DISABLE_BACKFACE_CULLING,
+    IS_BREAKABLE_STATUE,
+    IS_CRANE,
+};
+
```
Comment (Pirulax) [inline Client/sdk/game/CModelInfo.h]: Perhaps make this an `enum` and use it as a bit mask?
Or hack it like this, so nobody can access it like `IS_TAG`:
```cpp
namespace eModelIdeFlag
{
enum
{
    IS_ROAD,
    DRAW_LAST,
    ADDICTIVE,
    IGNORE_LIGHTING,            // Used with animated objects
    NO_ZBUFFER_WRITE,
    DONT_RECEIVE_SHADOWS,
    IS_GLASS_TYPE_1,
    IS_GLASS_TYPE_2,
    IS_GARAGE_DOOR,
    IS_DAMAGABLE,
    IS_TREE,
    IS_PALM,
    DOES_NOT_COLLIDE_WITH_FLYER,
    IS_TAG,
    DISABLE_BACKFACE_CULLING,
    IS_BREAKABLE_STATUE,
    IS_CRANE,
};
};
```

File: Client/game_sa/CModelInfoSA.h [hunk 12]
```diff
@@ -137,6 +139,20 @@ class CVehicleModelInfo_SA_VTBL : public CClumpModelInfo_SA_VTBL
...
+enum class eModelSpecialType : unsigned char
+{
+    NONE = 0,
+    TREE = 1,
+    PALM = 2,
+    GLASS_1 = 4,
+    GLASS_2 = 5,
+    TAG = 6,
+    GARAGE_DOOR = 7,
+    CRANE = 9, // don't present in IDE but used in gta
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.h]: Grammar:
> Not present in IDE but used in GTA

File: Client/game_sa/CModelInfoSA.h [hunk 13]
```diff
@@ -36,7 +36,9 @@ class CPedModelInfoSAInterface;
 #define     ARRAY_ModelInfo                 0xA9B0C8
 #define     CLASS_CText                     0xC1B340
 #define     FUNC_CText_Get                  0x6A0050
-#define     FUNC_GetModelFlags              0x4044E0
+#define     FUNC_SetBaseModelFlags          0x5B3AD0
+#define     FUNC_SetAtomicModelFlags        0x5B3B20
+#define     FUNC_GetModelStreamFlags        0x4044E0
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.h]: IIRC we've stopped using defines for function addresses.
@saml1er am I correct?

File: Client/game_sa/CModelInfoSA.cpp [hunk 14]
```diff
@@ -483,21 +521,278 @@ BOOL CModelInfoSA::DoIsLoaded()
...
+//    _asm
+//    {
+//        push    ModelID
+//        call    dwFunc
+//        add     esp, 4
+//        mov     bFlags, al
+//    }
+//    return bFlags;
+//}
+
```
Comment (Pirulax) [inline Client/game_sa/CModelInfoSA.cpp]: Left in by accident?

**Review by Pirulax — DISMISSED** [review]

File: Client/sdk/game/CModelInfo.h (hunk 11 above)
Comment (TheNormalnij) [reply Client/sdk/game/CModelInfo.h]: Yeah,  i have think about bit mask but i can't use it for some flags ( IS_DAMAGABLE, IS_TREE, IS_GLASS_TYPE_2, ... )

File: Client/game_sa/CModelInfoSA.cpp (hunk 9 above)
Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.cpp]: This is almost original GTA SA behavior. Game reads flags in the same way

File: Client/game_sa/CModelInfoSA.cpp (hunk 14 above)
Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.cpp]: Old code for GetFlags function but with wrong behavior

File: Client/game_sa/CModelInfoSA.h (hunk 13 above)
Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.h]: Hmm... Unused

File: Client/game_sa/CModelInfoSA.cpp (hunk 9 above)
Comment (Pirulax) [reply Client/game_sa/CModelInfoSA.cpp]: I dont know.. Its really weird, because `m_pInterface->eSpecialModelType` will always be overwritten by the last `true if`, thus a reversed `else if` ends up doing the same.
So you can just do a reversed `else if`:
```cpp

m_pInterface->bDontCollideWithFlyer = ideFlags.bDontCollideWithFlyer;

if (ideFlags.bFlag20)
    m_pInterface->eSpecialModelType = eModelSpecialType::CRANE;
else if (ideFlags.bFlag24)
    m_pInterface->eSpecialModelType = eModelSpecialType::UNKNOW_1;
else if (ideFlags.bIsBreakableStatue)
    m_pInterface->eSpecialModelType = eModelSpecialType::BREAKABLE_STATUE;
else if (ideFlags.bIsTag)
    m_pInterface->eSpecialModelType = eModelSpecialType::TAG;
else if (ideFlags.bIsPalm)
    m_pInterface->eSpecialModelType = eModelSpecialType::PALM;
else if (ideFlags.bIsTree)
    m_pInterface->eSpecialModelType = eModelSpecialType::TREE;
else if (ideFlags.bIsGarageDoor)
    m_pInterface->eSpecialModelType = eModelSpecialType::GARAGE_DOOR;
else if (ideFlags.bIsGlassType2)
    m_pInterface->eSpecialModelType = eModelSpecialType::GLASS_2;
else if (ideFlags.bIsGlassType1)
    m_pInterface->eSpecialModelType = eModelSpecialType::GLASS_1;
```

Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.cpp]: if i will use `else if` `m_pInterface->eSpecialModelType` will be overwritten with first `if`

Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.cpp]: Sounds good, thx

Comment (Pirulax) [reply Client/game_sa/CModelInfoSA.cpp]: You're welcome!

File: Client/game_sa/CModelInfoSA.cpp (hunk 7 above)
Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.cpp]: What this function can return?

Comment (Pirulax) [reply Client/game_sa/CModelInfoSA.cpp]: Is this function allowed to be called even if the interface for the model is `null`?
Because if not, you should just:
```cpp
unsigned short CModelInfoSA::GetFlags()
{
   m_pInterface = ppModelInfo[m_dwModelID];
   return m_pInterface->usFlags;
}
```
This will crash if `m_pInterface` is `null`, which is okay if this function shoudn't be called if `m_pInterface` is `null`.

File: Client/game_sa/CModelInfoSA.cpp (hunk 10 above)
Comment (qaisjp) [reply Client/game_sa/CModelInfoSA.cpp:750]: I don't see a problem with the code that is currently there

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.h [hunk 15]
```diff
@@ -57,10 +57,14 @@ class CLuaEngineDefs : public CLuaDefs
     LUA_DECLARE(EngineSetObjectGroupPhysicalProperty)
     LUA_DECLARE(EngineGetObjectGroupPhysicalProperty)
     LUA_DECLARE(EngineRestoreObjectGroupPhysicalProperties)
-    static bool CLuaEngineDefs::EngineRestreamWorld(lua_State* const luaVM);
+    static uint CLuaEngineDefs::EngineGetModelFlags(uint uiModelID);
+    static bool CLuaEngineDefs::EngineSetModelFlags(uint uiModelID, uint uiFlags, std::optional<bool> bIdeFlags);
+    static bool CLuaEngineDefs::EngineGetModelFlag(uint uiModelID, eModelIdeFlag eFlag);
+    static bool CLuaEngineDefs::EngineSetModelFlag(uint uiModelID, eModelIdeFlag eFlag, bool state);
+    static bool CLuaEngineDefs::EngineResetModelFlags(uint uiModelID);
+    static bool CLuaEngineDefs::EngineRestreamWorld();
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.h]: You don't need to have `CLuaEngineDefs::` here

File: Client/game_sa/CModelInfoSA.cpp [hunk 16]
```diff
@@ -19,12 +19,50 @@ CBaseModelInfoSAInterface** ppModelInfo = (CBaseModelInfoSAInterface**)ARRAY_Mod
...
 std::unordered_map<DWORD, std::pair<float, float>>                    CModelInfoSA::ms_VehicleModelDefaultWheelSizes;
 
+union tIdeFlags
+{
+    struct
+    {
+        char bIsRoad : 1;
+        char bFlag2 : 1;
+        char bDrawLast : 1;
+        char bAddictive : 1;
```
Comment (StrixG) [inline Client/game_sa/CModelInfoSA.cpp]: Typo
```suggestion
        char bAdditive : 1;
```

File: Client/game_sa/CModelInfoSA.cpp [hunk 17]
```diff
@@ -19,12 +19,50 @@ CBaseModelInfoSAInterface** ppModelInfo = (CBaseModelInfoSAInterface**)ARRAY_Mod
...
+        char bIsTree : 1;
+        char bIsPalm : 1;
+        char bDontCollideWithFlyer : 1;
+
+        char bFlag17 : 1;
+        char bFlag18 : 1;
+        char bFlag19 : 1;
+        char bFlag20 : 1;
+        char bIsTag : 1;
+        char bDisableBackfaceCilling : 1;
```
Comment (StrixG) [inline Client/game_sa/CModelInfoSA.cpp]:
```suggestion
        char bDisableBackfaceCulling : 1;
```

File: Client/game_sa/CModelInfoSA.cpp [hunk 18]
```diff
@@ -483,19 +521,252 @@ BOOL CModelInfoSA::DoIsLoaded()
...
+    {
+        case eModelIdeFlag::IS_ROAD:
+            m_pInterface->bIsRoad = bState;
+            break;
+        case eModelIdeFlag::DRAW_LAST:
+            m_pInterface->bAlphaTransparency = bState;
+            if (bState)
+                m_pInterface->bAdditiveRender = true;
+            break;
+        case eModelIdeFlag::ADDICTIVE:
```
Comment (StrixG) [inline Client/game_sa/CModelInfoSA.cpp]:
```suggestion
        case eModelIdeFlag::ADDITIVE:
```

File: Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp [hunk 19]
```diff
@@ -666,6 +666,26 @@ ADD_ENUM(eClientModelType::OBJECT, "object")
 ADD_ENUM(eClientModelType::VEHICLE, "vehicle")
 IMPLEMENT_ENUM_CLASS_END("client-model-type")
 
+IMPLEMENT_ENUM_CLASS_BEGIN(eModelIdeFlag)
+ADD_ENUM(eModelIdeFlag::IS_ROAD, "is_road")
+ADD_ENUM(eModelIdeFlag::DRAW_LAST, "draw_last")
+ADD_ENUM(eModelIdeFlag::ADDICTIVE, "addictive")
```
Comment (StrixG) [inline Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp]:
```suggestion
ADD_ENUM(eModelIdeFlag::ADDITIVE, "additive")
```

File: Client/game_sa/CModelInfoSA.cpp [hunk 20]
```diff
@@ -483,19 +521,252 @@ BOOL CModelInfoSA::DoIsLoaded()
...
+    if (!m_pInterface)
+        return false;
+
+    switch (eIdeFlag)
+    {
+        case eModelIdeFlag::IS_ROAD:
+            return m_pInterface->bIsRoad;
+        case eModelIdeFlag::DRAW_LAST:
+           return m_pInterface->bAlphaTransparency;
+        case eModelIdeFlag::ADDICTIVE:
```
Comment (StrixG) [inline Client/game_sa/CModelInfoSA.cpp]:
```suggestion
        case eModelIdeFlag::ADDITIVE:
```

File: Client/sdk/game/CModelInfo.h [hunk 21]
```diff
@@ -28,6 +28,27 @@ class CBoundingBox
     float   fRadius;
 };
 
+enum class eModelIdeFlag
+{
+    IS_ROAD,
+    DRAW_LAST,
+    ADDICTIVE,
```
Comment (StrixG) [inline Client/sdk/game/CModelInfo.h]:
```suggestion
    ADDITIVE,
```

File: Client/game_sa/CModelInfoSA.cpp (hunk 10 above)
Comment (TheNormalnij) [reply Client/game_sa/CModelInfoSA.cpp:718]: i think current state is more readable

**Comment by TheNormalnij** [comment]
conflicts solved, can be merged

**Comment by lopezloo** [comment]
Do all these flags actually work? `is_crane` for example. What does it do?

**Comment by TheNormalnij** [comment]
I saw almost all flags in IDA, but im not sure that all of them have effect in MTA. My "must have" list:
```
draw_last -- fix world models culling
addictive -- fix render issues
ignore_lighting
no_zbuffer_write
dont_receive_shadows
is_palm
```

**Comment by Shuubaru** [comment]
i'm curious, will this fix vehicle/peds shadows over non-world objects created with **createObject** function?

**Comment by Allerek** [comment]
Is there any docs about what these flags do? We can guess some for sure like
`is_palm` is probably about bending the object while wind is strong etc.

**Comment by lopezloo** [comment]
> i'm curious, will this fix vehicle/peds shadows over non-world objects created with **createObject** function?

It won't. MTA created objects are treated by game as small dynamic objects like lampposts etc. and that's why they do not reflect shadows.

**Comment by Lpsd** [comment]
As briefly mentioned on Discord - I think the API needs simplifying and the usage being more "consistent".

We should have only 2 functions for setting and getting flag(s). Each function you can either pass a single flag ID, or a table of flag IDs (ints).

Also, both functions should accept/return the flag IDs (ints). This is the consistency part, as currently you have to work with both ints as well as string names, depending on if you're setting one, or multiple flags.

We can have a table of the flag IDs and their names on the wiki.

This also gets rid of the need for that `bIdeFlags` arg.

Ideally, the functions would be:

```lua
table engineGetModelFlags( int modelID, int flagID / table flagIDs )
bool engineSetModelFlags( int modelID, int flagID / table flagIDs, bool state )
```

**Comment by TheNormalnij** [comment]
As i remember, GTA SA translates flags from IDE files in internal format, and we can't translate GTA format in IDE format.
I think map developers know better ide flags and will be discouraged with internal gta system.
Also people will use precalculated flags from ide files or MTA editors, and this is main usage, i think.

So, user friendly API should handle well-known information in this situation.

As Pirulax said once "No magic numbers in API". We have some unfriendly examples like `get/setVehicleWheelsSates`. I always check wiki about these functions. In my opinion `flagID` will be fast, but harder to use.

Current API is perfect for 2 types of operations:
1) You need set flags for a new object. You use `engineSetModelFlags` with your precalculated value or you get this value from `bitOr` with enabled flags from IDE or GTA.
2) You need tweak objects. If you need change only one flag, or disable/enable some flags for all models, `engineSetModelFlag` will help you.

**Comment by Zangomangu** [comment]
> As i remember, GTA SA translates flags from IDE files in internal format, and we can't translate GTA format in IDE format.

If we can't translate it back, I don't think we should expose it. Most people would set flags in the IDE format, because that's the documented format.
So you set a value, but you get a different value. That causes confusion. Is there any way we can just return the IDE formatted flags? GTA shouldn't be changing the model flags, so only this function would be altering them?


As pointed out above, current scripting functions take alternative parameters instead of having several functions. So we should keep it to one set/get/reset IMO.

I also agree we shouldn't come up with magic numbers, so what about the following.
Scripters not familiar with bitwise operations can use the string variant to set/get individual flags.

`int flags` for the IDE flags.
`string flagName` for a specific flag.

 ```lua
 int/bool engineGetModelFlags( int modelID [, string flagName ] )
 ```
Returns IDE flags, or state of a specific flag if flagName is specified.
Function should error if parsed an invalid modelID to avoid ambiguity.

 ```lua
 bool engineSetModelFlags( int modelID, int flags / string flagName [, bool state ] )
 ```
Sets IDE flags, or a specific flag if flagName and state is specified

**Comment by gta191977649** [comment]
> > As i remember, GTA SA translates flags from IDE files in internal format, and we can't translate GTA format in IDE format.
> 
> If we can't translate it back, I don't think we should expose it. Most people would set flags in the IDE format, because that's the documented format. So you set a value, but you get a different value. That causes confusion. Is there any way we can just return the IDE formatted flags? GTA shouldn't be changing the model flags, so only this function would be altering them?
> 
> As pointed out above, current scripting functions take alternative parameters instead of having several functions. So we should keep it to one set/get/reset IMO.
> 
> I also agree we shouldn't come up with magic numbers, so what about the following. Scripters not familiar with bitwise operations can use the string variant to set/get individual flags.
> 
> `int flags` for the IDE flags. `string flagName` for a specific flag.
> 
> ```lua
> int/bool engineGetModelFlags( int modelID [, string flagName ] )
> ```
> 
> Returns IDE flags, or state of a specific flag if flagName is specified. Function should error if parsed an invalid modelID to avoid ambiguity.
> 
> ```lua
> bool engineSetModelFlags( int modelID, int flags / string flagName [, bool state ] )
> ```
> 
> Sets IDE flags, or a specific flag if flagName and state is specified

I agreed, most people will use the flags id from corresponding idea file, this feature will benefit especially someone who are trying to do a map total convertion in MTA.

**Comment by patrikjuvonen** [comment]
Merge conflicts must be resolved.

**Review by patrikjuvonen — DISMISSED** [review]
This is great, thank you very much.

**Review by patrikjuvonen — DISMISSED** [review]
